File: frontend/src/services/announcement.ts
import type { Pending } from './project'
import axios from 'axios'

export interface AnnProject {
  id: number
  name: string
  number: string
  agency_code: string
  agency_name: string
  status: string
  round?: number
}

export interface AnnPortal {
  portal_status: string      // 挂网中|已挂网|挂网失败|撤网中|已撤网|官网上找不到
  portal_url: string         // 公网详情页地址
  portal_news_id: number     // 官网那条记录的 id
  portal_error: string
  portal_at?: string
  running?: boolean
  enabled?: boolean
}

export const getAnnPortal = (id: number) =>
  axios.get<{ ok: boolean; data: AnnPortal }>(`/api/announcements/${id}/portal`,
    { withCredentials: true })

export const publishAnnPortal = (id: number) =>
  axios.post<{ ok: boolean; message: string }>(`/api/announcements/${id}/portal/publish`, {},
    { withCredentials: true })

export const revokeAnnPortal = (id: number) =>
  axios.post<{ ok: boolean; message: string }>(`/api/announcements/${id}/portal/revoke`, {},
    { withCredentials: true })

export const recheckAnnPortal = (id: number) =>
  axios.post<{ ok: boolean; online: boolean; data: AnnPortal }>(
    `/api/announcements/${id}/portal/recheck`, {}, { withCredentials: true })

export interface Announcement {
  pending?: Pending          // 当前处理人
  id: number
  project_id: number
  project_name?: string
  project_number?: string
  agency_name?: string
  project_agency_code?: string
  ann_type: string
  ann_type_cn?: string
  round_number: number
  project_intro: string
  qualifications: string   // 一般资格要求（整段文本）
  special_req: string
  reg_start: string
  reg_end: string
  reg_note: string
  response_deadline: string
  agency_address: string
  delivery_address: string
  agency_email: string
  agency_reg_phone: string
  agency_contact: string
  agency_contact_phone: string
  // 更正公告（ann_type='correction'）专用
  // 医院官网挂网
  portal_status?: string
  portal_url?: string
  portal_news_id?: number
  portal_error?: string
  portal_at?: string
  corr_scope: string          // 更正事项：采购公告 / 采购文件 / 采购公告、采购文件
  corr_reason: string
  corr_items_json: string     // [{item,before,after}]
  corr_in_attachment: number  // 1=内容较多详见附件
  corr_seq: number            // 第几次更正
  // 6.2 调研公告
  survey_content?: string
  survey_qualification?: string
  survey_quote_req?: string
  survey_materials?: string
  survey_deadline?: string
  survey_submit_way?: string
  survey_note?: string
  // 6.4 单一来源公示（法定必备内容，财政部令第74号第38条）
  ss_goods_desc?: string
  ss_reason?: string
  ss_supplier_name?: string
  ss_supplier_addr?: string
  ss_experts_json?: string
  ss_publicity_start?: string
  ss_publicity_end?: string
  ss_objection_dept?: string
  ss_objection_contact?: string
  ss_objection_phone?: string
  ss_objection_addr?: string
  status: string              // 草稿|待确认|已确认|已驳回
  confirmed_by: string
  confirmed_at: string
  reject_reason?: string
  reject_count?: number
  rejected_by?: string
  rejected_at?: string
  created_at: string
  created_by: string
}

export interface AnnAttachment {
  id: number
  announcement_id: number
  original_name: string
  file_size: number
  uploaded_by: string
  uploaded_at: string
}

export const QUALIFICATIONS_DEFAULT =
  '（一）在中华人民共和国境内注册，具有独立法人资格；\n' +
  '（二）具有良好的商业信誉和健全的财务会计制度；\n' +
  '（三）具备履行合同所必需的设备和专业技术能力；\n' +
  '（四）参加采购活动前三年内，在经营活动中没有重大违法记录；\n' +
  '（五）本项目不接受联合体；\n' +
  '（六）本项目规定的其他要求：'

export type AnnFormData = Omit<Announcement,
  'id' | 'project_name' | 'project_number' | 'agency_name' | 'project_agency_code' |
  'ann_type_cn' | 'confirmed_by' | 'confirmed_at' | 'created_at' | 'created_by' | 'corr_seq'>

// ── 公开接口（无需登录） ────────────────────────────────────────
export function getPublicAnnouncements(annType = 'procurement') {
  return axios.get<{ ok: boolean; data: Announcement[] }>('/api/announcements/public', {
    params: { type: annType },
  })
}

export function getPublicAnnouncement(id: number) {
  return axios.get<{ ok: boolean; data: Announcement }>(`/api/announcements/public/${id}`)
}

export function getPublicAnnouncementFiles(annId: number) {
  return axios.get<{ ok: boolean; data: AnnAttachment[] }>(`/api/announcements/public/${annId}/files`)
}

export function publicDownloadFileUrl(annId: number, fileId: number) {
  return `/api/announcements/public/${annId}/files/${fileId}`
}

export function publicWordUrl(annId: number) {
  return `/api/announcements/public/${annId}/word`
}

export function getPublicAnnouncementHtml(annId: number) {
  return axios.get<{ ok: boolean; html: string }>(`/api/announcements/public/${annId}/html`)
}

// ── 公告 CRUD ──────────────────────────────────────────────────
export function getEligibleProjects(annType = 'procurement') {
  return axios.get<{ ok: boolean; data: AnnProject[] }>('/api/announcements/projects', {
    params: { type: annType },
  })
}
export function getAnnouncements(annType = 'procurement') {
  return axios.get<{ ok: boolean; data: Announcement[] }>('/api/announcements', {
    params: { type: annType },
  })
}
export function getAnnouncement(id: number) {
  return axios.get<{ ok: boolean; data: Announcement }>(`/api/announcements/${id}`)
}
export function createAnnouncement(data: AnnFormData) {
  return axios.post<{ ok: boolean; message: string; data: Announcement }>('/api/announcements', data)
}
export function updateAnnouncement(id: number, data: AnnFormData) {
  return axios.put<{ ok: boolean; message: string; data: Announcement }>(`/api/announcements/${id}`, data)
}
export function deleteAnnouncement(id: number) {
  return axios.delete<{ ok: boolean; message: string }>(`/api/announcements/${id}`)
}
export function submitAnnouncement(id: number) {
  return axios.post<{ ok: boolean; message: string; data: Announcement }>(`/api/announcements/${id}/submit`)
}
export function confirmAnnouncement(id: number) {
  return axios.post<{ ok: boolean; message: string; data: Announcement }>(`/api/announcements/${id}/confirm`)
}
export function revokeAnnouncement(id: number) {
  return axios.post<{ ok: boolean; message: string; data: Announcement }>(`/api/announcements/${id}/revoke`)
}
/** 驳回：打回代理机构修改，必须写明原因（记入审批过程记录，归档留存） */
export function rejectAnnouncement(id: number, reason: string) {
  return axios.post<{ ok: boolean; message: string; data: Announcement }>(
    `/api/announcements/${id}/reject`, { reason })
}
export function generateAnnouncementWord(id: number) {
  return axios.post(`/api/announcements/${id}/generate`, {}, { responseType: 'blob' })
}
// 点项目名在线预览生成的公告 Word（GET，内联）
export function announcementWordUrl(id: number) {
  return `/api/announcements/${id}/word`
}

// ── 附件 ──────────────────────────────────────────────────────
export function listFiles(annId: number) {
  return axios.get<{ ok: boolean; data: AnnAttachment[] }>(`/api/announcements/${annId}/files`)
}
export function deleteFile(annId: number, fileId: number) {
  return axios.delete<{ ok: boolean; message: string }>(`/api/announcements/${annId}/files/${fileId}`)
}
export function downloadFileUrl(annId: number, fileId: number) {
  return `/api/announcements/${annId}/files/${fileId}`
}
export function previewFileUrl(annId: number, fileId: number) {
  return `/api/announcements/${annId}/files/${fileId}/preview`
}

File: backend/models/announcement.py
from . import db

QUALIFICATIONS_DEFAULT = (
    '（一）在中华人民共和国境内注册，具有独立法人资格；\n'
    '（二）具有良好的商业信誉和健全的财务会计制度；\n'
    '（三）具备履行合同所必需的设备和专业技术能力；\n'
    '（四）参加采购活动前三年内，在经营活动中没有重大违法记录；\n'
    '（五）本项目不接受联合体；\n'
    '（六）本项目规定的其他要求：'
)


class Announcement(db.Model):
    __tablename__ = "announcements"

    id = db.Column(db.Integer, primary_key=True)
    project_id = db.Column(db.Integer, nullable=False)
    ann_type = db.Column(db.String(20), default="procurement")
    round_number = db.Column(db.Integer, default=1)

    # 正文内容
    project_intro = db.Column(db.Text, default="")
    qualifications = db.Column(db.Text, default=QUALIFICATIONS_DEFAULT)  # 一般资格要求（整段）
    special_req = db.Column(db.Text, default="")                          # 特殊要求（第七项）

    # 时间
    reg_start = db.Column(db.String(60), default="")
    reg_end = db.Column(db.String(60), default="")
    reg_note = db.Column(db.Text, default="")
    response_deadline = db.Column(db.String(60), default="")

    # 代理机构信息
    agency_address = db.Column(db.String(200), default="")
    delivery_address = db.Column(db.String(200), default="")
    agency_email = db.Column(db.String(100), default="")
    agency_reg_phone = db.Column(db.String(60), default="")
    agency_contact = db.Column(db.String(60), default="")
    agency_contact_phone = db.Column(db.String(60), default="")

    # ── 更正公告（ann_type='correction'）专用字段 ─────────────────
    corr_scope = db.Column(db.String(30), default="")        # 更正事项：采购公告 / 采购文件 / 采购公告、采购文件
    corr_reason = db.Column(db.Text, default="")             # 更正原因（简述）
    corr_items_json = db.Column(db.Text, default="[]")       # [{"item":..,"before":..,"after":..}]
    corr_in_attachment = db.Column(db.Integer, default=0)    # 1=更正内容较多，详见附件
    corr_seq = db.Column(db.Integer, default=1)              # 本项目本轮第几次更正（标题用）

    # ── 调研公告（ann_type='survey'）专用字段 ──────────────────────
    # 体例照内江一院官网已发布的市场调研公告（如云算力、法律顾问两份）
    survey_content       = db.Column(db.Text, default="")    # 一、调研内容及技术/服务要求
    survey_qualification = db.Column(db.Text, default="")    # 二、供应商资格要求
    survey_quote_req     = db.Column(db.Text, default="")    # 三、报价要求
    survey_materials     = db.Column(db.Text, default="")    # 四、需提交的资料清单
    survey_deadline      = db.Column(db.String(60), default="")   # 提交截止时间
    survey_submit_way    = db.Column(db.Text, default="")    # 提交方式（邮箱/纸质地址）
    survey_note          = db.Column(db.Text, default="")    # 特别说明（默认含"与采购结果无必然联系"）

    # ── 单一来源公示（ann_type='single_source'）专用字段 ────────────
    # 必备内容按《政府采购非招标采购方式管理办法》（财政部令第74号）第38条
    ss_goods_desc     = db.Column(db.Text, default="")       # 拟采购的货物或服务说明
    ss_reason         = db.Column(db.Text, default="")       # 采用单一来源方式的原因及说明
    ss_supplier_name  = db.Column(db.String(200), default="")  # 拟定唯一供应商名称
    ss_supplier_addr  = db.Column(db.String(300), default="")  # 唯一供应商地址
    ss_experts_json   = db.Column(db.Text, default="[]")     # [{name,org,title,opinion}] 专家论证意见
    ss_publicity_start = db.Column(db.String(60), default="")  # 公示期起（法定不少于5个工作日）
    ss_publicity_end   = db.Column(db.String(60), default="")  # 公示期止
    ss_objection_dept  = db.Column(db.String(120), default="")  # 异议接收部门
    ss_objection_contact = db.Column(db.String(60), default="")
    ss_objection_phone   = db.Column(db.String(60), default="")
    ss_objection_addr    = db.Column(db.String(300), default="")

    # 状态：草稿 → 待确认 →（经办人）已确认 / 已驳回 →（代理改后重提）待确认
    status = db.Column(db.String(20), default="草稿")
    confirmed_by = db.Column(db.String(50), default="")
    confirmed_at = db.Column(db.String(30), default="")
    # 驳回：原因保留在单据上供代理机构直接看到，完整往返链条在 approval_logs
    reject_reason = db.Column(db.Text, default="")
    reject_count = db.Column(db.Integer, default=0)
    rejected_by = db.Column(db.String(50), default="")
    rejected_at = db.Column(db.String(30), default="")

    created_at = db.Column(db.String(30), default="")
    created_by = db.Column(db.String(50), default="")

    def to_dict(self):
        return {
            "id": self.id,
            "project_id": self.project_id,
            "ann_type": self.ann_type,
            "round_number": self.round_number,
            "project_intro": self.project_intro or "",
            "qualifications": self.qualifications or QUALIFICATIONS_DEFAULT,
            "special_req": self.special_req or "",
            "reg_start": self.reg_start or "",
            "reg_end": self.reg_end or "",
            "reg_note": self.reg_note or "",
            "response_deadline": self.response_deadline or "",
            "agency_address": self.agency_address or "",
            "delivery_address": self.delivery_address or "",
            "agency_email": self.agency_email or "",
            "agency_reg_phone": self.agency_reg_phone or "",
            "agency_contact": self.agency_contact or "",
            "agency_contact_phone": self.agency_contact_phone or "",
            "corr_scope": self.corr_scope or "",
            "corr_reason": self.corr_reason or "",
            "corr_items_json": self.corr_items_json or "[]",
            "corr_in_attachment": int(self.corr_in_attachment or 0),
            "corr_seq": int(self.corr_seq or 1),
            "survey_content": self.survey_content or "",
            "survey_qualification": self.survey_qualification or "",
            "survey_quote_req": self.survey_quote_req or "",
            "survey_materials": self.survey_materials or "",
            "survey_deadline": self.survey_deadline or "",
            "survey_submit_way": self.survey_submit_way or "",
            "survey_note": self.survey_note or "",
            "ss_goods_desc": self.ss_goods_desc or "",
            "ss_reason": self.ss_reason or "",
            "ss_supplier_name": self.ss_supplier_name or "",
            "ss_supplier_addr": self.ss_supplier_addr or "",
            "ss_experts_json": self.ss_experts_json or "[]",
            "ss_publicity_start": self.ss_publicity_start or "",
            "ss_publicity_end": self.ss_publicity_end or "",
            "ss_objection_dept": self.ss_objection_dept or "",
            "ss_objection_contact": self.ss_objection_contact or "",
            "ss_objection_phone": self.ss_objection_phone or "",
            "ss_objection_addr": self.ss_objection_addr or "",
            "status": self.status or "草稿",
            "confirmed_by": self.confirmed_by or "",
            "confirmed_at": self.confirmed_at or "",
            "reject_reason": self.reject_reason or "",
            "reject_count": int(self.reject_count or 0),
            "rejected_by": self.rejected_by or "",
            "rejected_at": self.rejected_at or "",
            "created_at": self.created_at or "",
            "created_by": self.created_by or "",
        }
